### PR TITLE
[Fix for the GitLab CI] phist: prefer @1.9.5 (1.9.6 is not compatible w/ mpich%gcc:9)

### PR DESCRIPTION
**Pull request by bernhardkaindl**
phist is a required part of spack GitLab CI pipelines. In them, `mpich` is requested to provide 'mpi' like this:
```py
spack install phist ^mpich %gcc@7.5.0
```
Failure of this command to succeed breaks spack's gitlab CI pipelines.

phist-1.9.6 updated from the older "use mpi" to the newer "use mpi_f08" (MPI3.1):
* The motivation was fixing it on Cray: https://github.com/spack/spack/issues/26002
* Documentation: https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node408.htm

But, mpich does not provide mpi_f08.mod with gfortran-[789], it needs gfortran>=10: https://stackoverflow.com/questions/65750862

Tis means, as long as spack GitLab CI pipelines use `^mpich %gcc@7.5.0`, we cannot use `phist@1.9.6` in them.

I personlally guess moving the gitlab pipelines and E4S to **openmpi** would be the best solution, which is the default also, but I don't know if that is possible.

### Note:

The tests of all `phist` versions run into timeouts when **`^mpich`** is used. With **`^openmpi`**, they work.

I added all needed ingredents to run them with **`spack install --test=root`**, with the previous PR:

This means that the test timeouts with **`^mpich`** are easy to confirm and reproduce now.

### CODE REVIEW

File: var/spack/repos/builtin/packages/phist/package.py
Comment (tgamblin) [inline var/spack/repos/builtin/packages/phist/package.py]: I think this will be easier to model when we have compilers as dependencies (our next big thing to work on after 0.17).

This basically means that 1.9.6 all of a sudden requires Fortran 2008, at least with mpich, which is a large (probably inadvertent) change for a patch version 😬.  That's really a requirement on the compiler used, though in this case it also seems like it's a function of the particular bindings used by mpich: https://lists.mpich.org/pipermail/discuss/2018-March/002203.html.  It looks like gcc 7.5.0 should be able to support it, but doesn't.

I think the right thing to do here is not to delete 1.9.6, but to make 1.9.5 preferred:

```suggestion
    # The safest change (that still allows people to install 1.9.6 if they want) is to
    # prefer 1.9.5 for now. To verify future PRs, this command needs to be verified
    # to work for PRs:
    # spack install phist ^mpich %gcc@7.5.0
    version('1.9.6', sha256='98ed5ccb22bb98d5b6bf9de0c9960105473e5244978853070b9a3c44138db662')
    version(
        '1.9.5',
        sha256='24faa3373003f185c82a658c510e36cba9acc4110eb60cbfded9de370ae9ea32',
        preferred=True,
    )
```

What do you think?

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/phist/package.py
Comment (sethrj) [reply var/spack/repos/builtin/packages/phist/package.py]: Yes definitely prefer the older.

**Comment by sethrj** [comment]
Funny, I caught the mpich error in the E4S Phist pipeline at https://github.com/spack/spack/pull/26540 , but that one was closed because I guess they decided to tag a new revision than try to release new software under the same version? I didn't realize the newer 1.9.6 was trying to do the same thing.

File: var/spack/repos/builtin/packages/phist/package.py
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/phist/package.py]: @tgamblin @wspear @jthies I've added `preferred=True` for 1.9.5, keeping 1.9.6 available as you requested.

Further fixes:
* As pointed out by @sethrj in https://github.com/spack/spack/issues/26002#issuecomment-932952317 `-fallow-argument-mismatch` is really needed to fix the build with `%gcc@10:` - added:
```py
+        # Workaround for non-compliant and possibly broken code needed for gcc@10:
+        # MPI_Allreduce(localRes, res, 1, MPI_LOGICAL, MPI_LAND, map1%comm, ierr);
+        # MPI_Allreduce(my_ierr,global_ierr,1,MPI_INTEGER,MPI_MAX,map1%comm,mpi_ierr)
+        # Error: Type mismatch between actual argument at (1)
+        # and actual argument at (2) (LOGICAL(4)/INTEGER(4)).
+
+        if spec.satisfies('%gcc@10:'):
+            args.append(define('CMAKE_Fortran_FLAGS', '-fallow-argument-mismatch'))
```

* As @sethrj also pointed out, the git tag 1.9.5 from which we get the tarball has been set to an older commit. It does not have the correct version, so to give the correct version, VERSION_PATCH has to be incremented from 4 (1.9.4) to 5 (1.9.5):
```py
+            # Tag '1.9.5' has moved to an older commit later without fixing the version:
+            filter_file('VERSION_PATCH 4', 'VERSION_PATCH 5', 'CMakeLists.txt')
```
* Parallel build only works since version 1.8.0 (1.7.x failed to compile because of this)
  - added `@propery  def parallel(self): return self.spec.satisfies('@1.8:')` to set it accordingly.

* With `^mpich` all version and gcc compiler combinations I the dual and triple-process (np2 and np3) tests always timed-out, not even 1000 seconds were enough, at least not on a 24-core/48-thread AMD Rome CPU.
  - I added a hint to tell that the tests work with ^openmpi if the tests fail when ^mpi is selected.

**Review by sethrj — DISMISSED** [review]
Looks great, a shame it's our job to repeatedly patch this software.

**Comment by sethrj** [comment]
@bernhardkaindl Looks like the phist version for the compiler is still concretizing to 1.9.6, so [the CI is still failing](https://gitlab.spack.io/spack/spack/-/jobs/1146384):
```
==> [2021-10-16-12:37:25.768303] InstallError: PR 26773: gcc<10 can't build phist>1.9.5 w/ ^mpich
==> [2021-10-16-12:37:25.793338] Error: InstallError: PR 26773: gcc<10 can't build phist>1.9.5 w/ ^mpich
```

File: var/spack/repos/builtin/packages/phist/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/phist/package.py:60]: Maybe the preferred is being ignored because of the version 1.9.5 immediately below?

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/phist/package.py:60]: Thanks! Yes that's the cause why 1.9.6 was still the default.
This fixed it for `spack install phist`. Lets hope it fixes for the CI too.

**Review by sethrj — APPROVED** [review]
Will need @tgamblin 's signoff as well.

**Comment by jthies** [comment]
It's kind of hard to satisfy every possible build combination, the use of mpi_f08 instead of the mpi module was meant to make things more safe and portable, but then selecting an old compiler/mpi combination again seems to break it. Would adding a conflict('^mpich %gcc@7.5.0' when='@1.9.6:') be an option? But then you can't have that case in your CI, obviously. Maybe clingo would automatically switch to 1.9.5?

**Review by tgamblin — APPROVED** [review]

**Comment by tgamblin** [comment]
I think the range of incompatibilities is broader than just gcc@7.5.0. Probably the right thing to do eventually is to express this in terms of required f08 features, but we don’t have that granularity (yet). Let’s revisit when we do.

**Comment by bernhardkaindl** [comment]
Answer for @jthies:
> Would adding a conflict('^mpich %gcc@7.5.0' when='@1.9.6:') be an option? But then you can't have that case in your CI, obviously. Maybe clingo would automatically switch to 1.9.5?

No, I tried adding a similar conflict, and the result was that `spack install -v --test=root phist@1.9.6 %gcc@7.5.0` caused clingo to creat a completely broken concretization, in part using gcc@11 and in part gcc@7. We'll have to look at this in different scope.

I proposed adding the conflict (before I tested it) as well, and for now the (wise) decision of the core spack maintainers is to prefer 1.9.5 instead of the conflict. But all this is besides the point, this is just the fallout which spack has to deal with.

Your reign as maintainer of phist is a different one, a you are creating the software.

I want to come back to your area, your reign:

> It's kind of hard to satisfy every possible build combination, the use of mpi_f08 instead of the mpi module was meant to make things more safe and portable, but then selecting an old compiler/mpi combination again seems to break it.

I know. The many variants become manageable when you have a test script which tests all supported/tested combinations.

Of course such test has to draw the line somewhere because if you'd have 8 on/off variant's and all combinations would be possible, then 256 builds would have to be done, so you have to decide what needs to work and what should just work fine.

I've read that the trigger for enabling `mpi_f08` was Issue #26002, where this error was reported by wspear:
```py
Error: Type mismatch between actual argument at (1) and actual argument at (2) (LOGICAL(4)/INTEGER(4)).
```
Seth explained that this error *is* caused by the two mismatching calls to MPI_Allreduce: https://github.com/spack/spack/issues/26002#issuecomment-932952317 https://github.com/spack/spack/issues/26002#issuecomment-933041972
```fortran
grep -n MPI_Allreduce essex-phist-487ac8cd6198/src/kernels/builtin/map_module.F90
168:          call MPI_Allreduce(localRes, res, 1, MPI_LOGICAL, MPI_LAND, map1%comm, ierr);
391:      call MPI_Allreduce(my_ierr,global_ierr,1,MPI_INTEGER,MPI_MAX,map1%comm,mpi_ierr)
```
According to the gfortran developer's stance, in a bug report to gcc, this is not compliant and possibly buggy, which which was the reason why the gfortran developers turned this into an error with gfortan-10, which you can downgrade to a warning with `-fallow-argument-mismatch`, but they intentionally reject adding a way to disabling the warning message itself because of this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=91556#c24
Quoting from this bugzilla comment (empasis mine - they cannot ignore such mismatches while enabling LTO support in gfortran):
> > Also, this sort of code is a major obstacle for LTO, If we do not fix this, then I might as well give up on making gfortran LTO clean. Is the current treatment only necessary for LTO?
>
> Of course.  **Thomas quoted a part of the standard that indicate the code is invalid Fortran. I quoted a different line from the standard that indicates the code is invalid Fortran.  gfortran is informing the user the code is broken.**
>
>> Otherwise, can procedures without explicit interfaces be marked that they are not eligible to this kind of optimization?
>
>It seems that you missed an important aspect of Thomas's patch.  It allows one to find suspect subprogram references when the subprograms are implicitly defined.  Go read the first 6 comments in the PR. Jurgen had a bug in his code.  Thomas's patch generated an error pointing to the problem.  **With Jurgen's code, there is a big difference in passing a double precision value as opposed to the desired quad precision value.**
>
> *I'm having a hard time understanding why you think gfortran should continue to silently compile invalid, and possibly broken, code instead of actually alerting a user of an issue*.  I'm having an even harder time understanding why you want to have the ability to completely disable the

I know that the clingo solver made things more difficult by not telling you that it had to use gcc-9, despite you asking it to use %gcc@11.2.0 with the `spack spec phist %gcc@11.2.0 ^cmake/jcjw6cc` in this comment:
https://github.com/spack/spack/issues/26002#issuecomment-926502444

I assume this behavior must have led you to the (obviously wrong) assumption that `spack` just compiled `phist` with gcc-10 or gcc-11, while in fact it didn't.

The reality is that "out of the box", without using `cmake -D CMAKE_Fortran_FLAGS=-fallow-argument-mismatch`, gfortran >= 10 do issue the `Type mismatch between actual argument at (1) and actual argument at (2) (LOGICAL(4)/INTEGER(4))` error.

I reproduced this as soon as tried compile phist with %gcc@10 and @11 (in my case, I didn't select a build of cmake that was built with gcc@9, those compilers were really used, not not gcc@9).

Besides using %gcc@10.3.0 in the Cray environment of PR#26002 @wspear's Cray environment used an cray-specific mpich, but since that appears to have been compiled by gcc-9.1, it would likely not even have `mpi_f08`, because for it, upstream mpich needs gcc-10 or newer (as is explained by the link in the description of this PR).

If melven is correct with his comment https://github.com/spack/spack/issues/26002#issuecomment-925774683:
> The output and environment file indicate that the MPI library was built for a different compiler version (9.1 vs. 10.3).
I don't know how the details are handled but usually the generated MPI Fortran module file (mpi.mod) is not compatible between different compiler versions.

... then the correct response would have been t wait what the test programs posted by melven (from the linked comment) and of Seth from https://github.com/spack/spack/issues/26002#issuecomment-933041972 do in @wspear's Cray environment.

Seth and now I too bet on that especially Seth's test program won't compile in gcc-10-based Cray system without `-fallow-argument-mismatch` because the error from @wspear already said so.

Thus while more modern, using `mpi_f08` is likely not possible as that module is likey not provided by Cray, unless Cray also has an mpich build which is based on gfortan-10 or gfortan-11, and not only gcc-9 (which does not provide the extension mpich needs for mpi_f08).

I therefore think just reverting the two commits which started the migration to mpi_f08, together with -fallow-argument-mismatch makes things compile again (as long as a compatible mpich is used):

https://bitbucket.org/essex/phist/commits/585200643186f1825ea0af31d39bbbc4eb0a3b6e
https://bitbucket.org/essex/phist/commits/92fda4b1bc8a11d436980c756f121fbb29b4df64

If @wspear's `/opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/lib` is indeed compiled with gcc-9.1 and not compatible with the gcc-10.3.0 he tried to use, he'd have to use %gcc@9 to compile phist, and for that he likely needs an `phist` which does not depend on `mpi_f08`. Therefore just revert these 2 commits, and let him test `spack install phist@develop`

**Comment by bernhardkaindl** [comment]
> I think the range of incompatibilities is broader than just gcc@7.5.0. Probably the right thing to do eventually is to express this in terms of required f08 features, but we don’t have that granularity (yet). Let’s revisit when we do.

Yes, I've started discussion #26776 for do visit this.

Summarizing my previous comment on a single line:
When `phist` reverts to just `use mpi` again, we'd be able to use future `phist` versions in the e4s/mpich GitLab CI pipelines.

After this, the mismatching calls to MPI_Allreduce in src/kernels/builtin/map_module.F90 might be ideal to be fixed to match if possible. Then `-fallow-argument-mismatch` would not be needed for gfortan-10 and link-time-optimisations(LTO/-fto) might have better chances to be working: https://github.com/spack/spack/issues/26002#issuecomment-932952317 https://github.com/spack/spack/issues/26002#issuecomment-933041972
```fortran
grep -n MPI_Allreduce essex-phist-487ac8cd6198/src/kernels/builtin/map_module.F90
168:          call MPI_Allreduce(localRes, res, 1, MPI_LOGICAL, MPI_LAND, map1%comm, ierr);
391:      call MPI_Allreduce(my_ierr,global_ierr,1,MPI_INTEGER,MPI_MAX,map1%comm,mpi_ierr)
```

**Comment by sethrj** [comment]
Well said @bernhardkaindl , I think you hit the nail on the head. It's likely that since Cray's MPI module apparently provides only a single signature for the multiple Fortran overloads, it will not provide the f08 module and the update will still fail on cray.

In the meantime, the "allow mismatch" flag has been the path forward for other projects and is a suitable short term fix for phist . The use of the f08 seems to be causing more problems than it solves.

**Pull request merged**
